BlueDBOnDisk/src/main/java/io/bluedb/disk/segment/Segment.java

```diff
@@ -7,88 +7,129 @@
 import io.bluedb.disk.Blutils;
 import io.bluedb.disk.serialization.BlueSerializer;
 
 public class Segment <T extends Serializable> {
-	private static String SUFFIX = ".segment";
-
-	private final String pathString;
+//	private final String pathString;
 	private final Path path;
 	private final BlueSerializer serializer;
 	
-	public Segment(Path collectionPath, String segmentId, BlueSerializer serializer) {
-		this.path = Paths.get(collectionPath.toString(), segmentId + SUFFIX);
-		this.pathString = this.path.toString();
-		this.serializer = serializer;
-	}
-
-	public Segment(Path collectionPath, long segmentId, BlueSerializer serializer) {
-		this.path = Paths.get(collectionPath.toString(), segmentId + SUFFIX);
-		this.pathString = this.path.toString();
+	public Segment(Path segmentPath, BlueSerializer serializer) {
+		this.path = segmentPath;
+//		this.pathString = this.path.toString();
 		this.serializer = serializer;
 	}
+//
+//	public Segment(Path collectionPath, String segmentId, BlueSerializer serializer) {
+//		this(Paths.get(collectionPath.toString(), segmentId), serializer);
+//	}
+//
+//	public Segment(Path collectionPath, long segmentId, BlueSerializer serializer) {
+//		this(collectionPath, String.valueOf(segmentId), serializer);
+//	}
 
 	public void put(BlueKey key, T value) throws BlueDbException {
-		TreeMap<BlueKey, BlueEntity<T>> data = load();
+		TreeMap<BlueKey, BlueEntity<T>> data = load(key.getGroupingNumber());
 		BlueEntity<T> entity = new BlueEntity<T>(key, value);
 		data.put(key, entity);
+		String pathString = Paths.get(path.toString(), String.valueOf(key.getGroupingNumber())).toString();
 		Blutils.save(pathString, data, serializer);
 	}
 
 	public void delete(BlueKey key) throws BlueDbException {
-		TreeMap<BlueKey, BlueEntity<T>> data = load();
+		TreeMap<BlueKey, BlueEntity<T>> data = load(key.getGroupingNumber());
 		data.remove(key);
+		String pathString = Paths.get(path.toString(), String.valueOf(key.getGroupingNumber())).toString();
 		Blutils.save(pathString, data, serializer);
 	}
 
 	public BlueEntity<T> read(BlueKey key) throws BlueDbException {
-		Collection<BlueEntity<T>> values = load().values();
+		Collection<BlueEntity<T>> values = load(key.getGroupingNumber()).values();
 		for (BlueEntity<T> entity: values) {
 			if (entity.getKey().equals(key))
 				return entity;
 		}
 		return null;
 	}
 
-	public List<BlueEntity<T>> read() throws BlueDbException {
-		List<BlueEntity<T>> results = new ArrayList<>(load().values());
-		return results;
-	}
-
 	public List<BlueEntity<T>> read(long minTime, long maxTime) throws BlueDbException {
 		List<BlueEntity<T>> results = new ArrayList<>();
-		for (BlueEntity<T>entity: load().values()) {
+		List<Long> groupingNumbers = getAllFiles().stream()
+				.filter((f) -> isFileNameALong(f))
+				.map((f) -> f.getName())
+				.map((filename) -> Long.valueOf(filename))
+				.collect(Collectors.toList());
+		for (long groupingNumber: groupingNumbers) {

```

Comment:
You could choose to do a forEach here instead of collecting then iterating over the list again. A little more efficient though less clean.